File: solutions/java/triangle/2/src/main/java/Triangle.java
record Triangle(double a, double b, double c) {
  Triangle {
    if (a + b <= c || a + c <= b || b + c <= a) throw new TriangleException();
  }

  boolean isEquilateral() {
    return a == b && b == c;
  }

  boolean isIsosceles() {
    return a == b || a == c || b == c;
  }

  boolean isScalene() {
    return !isIsosceles();
  }
}

File: solutions/kotlin/triangle/2/src/main/kotlin/Triangle.kt
class Triangle<out T : Number>(val a: T, val b: T, val c: T) {

    init {
        listOf(a, b, c)
            .map(Number::toDouble)
            .apply { require(all { it > 0 }) { "All sides must be positive" } }
            .apply { require(sum() > 2 * max()) { "Triangle inequality violated" } }
    }

    val isEquilateral: Boolean get() = a == b && b == c
    val isIsosceles: Boolean get() = a == b || b == c || a == c
    val isScalene: Boolean get() = !isIsosceles
}
